File: appbiomedica/Modulos/Sesion/MenuItems.go
package Sesion

type Permits struct {
	Permiso string `json:"permiso,omitempty"`
}
type SubChildren struct {
	State  string `json:"state,omitempty"`
	Name   string `json:"name,omitempty"`
	Type   string `json:"type,omitempty"`
	Permit string `json:"permit,omitempty"`
}
type ChildrenItems struct {
	State  string        `json:"state,omitempty"`
	Name   string        `json:"name,omitempty"`
	Type   string        `json:"type,omitempty"`
	Permit string        `json:"permit,omitempty"`
	Child  []SubChildren `json:"child,omitempty"`
}
type Menu struct {
	State     string          `json:"state,omitempty"`
	Name      string          `json:"name,omitempty"`
	Type      string          `json:"type,omitempty"`
	Icon      string          `json:"icon,omitempty"`
	Permit    string          `json:"permit,omitempty"`
	Children  []ChildrenItems `json:"children,omitempty"`
}

var MENUITEMS = []Menu{
	{
		State:  "inicio",
		Name:   "Inicio",
		Type:   "link",
		Icon:   "apps",
		Permit: "Consulta",
	},
	{
		State:  "catalogos",
		Name:   "Catálogos",
		Type:   "sub",
		Icon:   "assignment",
		Permit: "Consulta",
		Children: []ChildrenItems{
			{State: "sai", Name: "SAI", Type: "link"},
			{State: "prei", Name: "PREI", Type: "link"},
			{State: "proforma", Name: "PROFORMA", Type: "link"},
		},
	},
	{
		State:  "listas",
		Name:   "Sitios de Interés",
		Type:   "link",
		Icon:   "link",
		Permit: "Consulta",
	},
	{
		State:  "area",
		Name:   "Areas Servicios",
		Type:   "sub",
		Icon:   "business",
		Permit: "Enfermeras",
		Children: []ChildrenItems{
			{State: "nueva", Name: "Agregar", Type: "link"},
		},
	},
	{
		State:  "planeacion",
		Name:   "Fondo Fijo",
		Type:   "sub",
		Icon:   "today",
		Permit: "Enfermeras",
		Children: []ChildrenItems{
			{State: "enfermeria", Name: "Enfermeria", Type: "subchild", Child: []SubChildren{
				{State: "cuadrosi", Name: "En Cuadro Basico", Type: "link"},
				{State: "cuadrono", Name: "Fuera de Cuadro Basico", Type: "link"},
			},
			},
		},
	},
	{
		State:  "solicitudes",
		Name:   "Solicitudes",
		Type:   "sub",
		Icon:   "description",
		Permit: "Solicitud",
		Children: []ChildrenItems{
			{State: "nueva", Name: "Nueva", Type: "subchild", Child: []SubChildren{
				{State: "consumo", Name: "Bienes de Consumo", Type: "link"},},
			},
			{State: "seguimiento", Name: "Seguimiento", Type: "link"},
			{State: "editar", Name: "Editar", Type: "link"},
		},
	},
	{
		State:  "configuraciones",
		Name:   "Configuraciones",
		Type:   "sub",
		Icon:   "build",
		Permit: "Jefe Biomedico",
		Children: []ChildrenItems{
			{State: "fecha", Name: "Fecha Límite", Type: "link"},
			{State: "presupuesto", Name: "Presupuesto", Type: "link"},
		},
	},
	{
		State:  "usuarios",
		Name:   "Usuarios",
		Type:   "sub",
		Icon:   "supervised_user_circle",
		Permit: "Jefe Biomedico",
		Children: []ChildrenItems{
			{State: "validar", Name: "Validar Usuarios", Type: "link"},
			{State: "bloquear", Name: "Bloquear", Type: "link"},
		},
	},
	{
		State:  "requerimientos",
		Name:   "Validar Solicitudes",
		Type:   "sub",
		Icon:   "assignment_turned_in",
		Permit: "Jefe Biomedico",
		Children: []ChildrenItems{
			{State: "revision", Name: "Validar Solicitud", Type: "link"},
			{State: "validado", Name: "Solicitudes Aprobadas", Type: "link"},
			{State: "final", Name: "Solicitud Final", Type: "link"},
		},
	},
	{
		State:  "proveedores",
		Name:   "Proveedores",
		Type:   "sub",
		Icon:   "apartment",
		Permit: "Consulta",
		Children: []ChildrenItems{
			{State: "lista", Name: "Lista de Proveedores", Type: "link"},
		},
	},
	{
		State:  "contratos",
		Name:   "Contratos",
		Type:   "sub",
		Icon:   "receipt",
		Permit: "Administrador de contratos",
		Children: []ChildrenItems{
			{State: "administrar", Name: "Administrar", Type: "link"},
			{State: "seguimiento", Name: "Seguimiento", Type: "link"},
		},
	},
	{
		State:  "traslado",
		Name:   "Traslados de pacientes",
		Type:   "sub",
		Icon:   "receipt",
		Permit: "Solicitud traslado",
		Children: []ChildrenItems{
			{State: "solicitud", Name: "Nueva Solicitud", Type: "link"},
			{State: "realizadas", Name: "Solicitudes Realizadas", Type: "link"},
			{State: "cancelar", Name: "Cancelar Solicitud", Type: "link"},
		},
	},
	{
		State:  "traslado",
		Name:   "Traslados de pacientes",
		Type:   "sub",
		Icon:   "receipt",
		Permit: "Administrador Traslado",
		Children: []ChildrenItems{
			{State: "firma", Name: "Firmas", Type: "link"},
			{State: "solicitud", Name: "Nueva Solicitud", Type: "link"},
			{State: "realizadas", Name: "Solicitudes Realizadas", Type: "link"},
			{State: "cancelar", Name: "Cancelar Solicitud", Type: "link"},
			{State: "manager", Name: "Administrar Traslados", Type: "link"},
		},
	},
}

/*****  Funcion para filtrar el menu item segun los permisos del usuario  *****/
func FiltrarPermisos(permisos []Permits) (error, []Menu) {
	var menuItem []Menu
	for _, permis := range permisos {
		for _, item := range MENUITEMS {
			if permis.Permiso == item.Permit {
				menuItem = append(menuItem, item)
			}
		}
	}
	return nil, menuItem
}

File: front/src/app/shared/menu-items/menu-items.ts
import {Injectable} from '@angular/core';


export interface BadgeItem {
  type: string;
  value: string;
}

export interface Separator {
  name: string;
  type?: string;
}

export interface SubChildren {
  state: string;
  name: string;
  type?: string;
  permit?: string;
}

export interface ChildrenItems {
  state: string;
  name: string;
  type?: string;
  child?: SubChildren[];
}

export interface Menu {
  state: string;
  id:string;
  moduloId:string;
  orden:number;
  name: string;
  type: string;
  icon: string;
  status:boolean;
  permit?: string;
  badge?: BadgeItem[];
  separator?: Separator[];
  children?: ChildrenItems[];
}

// const MENUITEMS = [
//   {
//     state: 'inicio',
//     name: 'Inicio',
//     type: 'link',
//     icon: 'apps',
//     permiso: 'Consulta',
//   },
//   {
//     state: 'catalogos',
//     name: 'Catálogos',
//     type: 'sub',
//     icon: 'assignment',
//     permiso: 'Consulta',
//     children: [
//       {state: 'sai', name: 'SAI', type: 'link'},
//       {state: 'prei', name: 'PREI', type: 'link'},
//       {state: 'proforma', name: 'PROFORMA', type: 'link'},
//       /*{state: 'manuales', name: 'MANUALES', type: 'link'},
//       {state: 'otros', name: 'OTROS CATALOGOS', type: 'link'}*/
//
//     ]
//   },
//   {
//     state: 'listas',
//     name: 'Listas de Interés',
//     type: 'link',
//     icon: 'link',
//     permiso: 'Consulta'
//   },
//   {
//     state: 'area',
//     name: 'Areas Servicios',
//     type: 'sub',
//     icon: 'business',
//     permiso: 'Enfermeras',
//     children: [
//       {state: 'nueva', name: 'Agregar', type: 'link'},
//     ]
//   },
//   {
//     state: 'planeacion',
//     name: 'Fondo Fijo',
//     type: 'sub',
//     icon: 'today',
//     permiso: 'Enfermeras',
//
//     children: [
//       {
//         state: 'enfermeria', name: 'Enfermeria', type: 'subchild',
//         subchildren: [
//           {
//             state: 'cuadrosi',
//             name: 'En Cuadro Basico',
//             type: 'link'
//           },
//           {
//             state: 'cuadrono',
//             name: 'Fuera de Cuadro Basico',
//             type: 'link'
//           },
//         ]
//       },
//     ]
//   },
//   {
//     state: 'solicitudes',
//     name: 'Solicitudes',
//     type: 'sub',
//     icon: 'description',
//     permiso: 'Solicitud',
//
//     children: [
//       {
//         state: 'nueva',
//         name: 'Nueva',
//         type: 'subchild',
//         subchildren: [
//           {
//             state: 'consumo',
//             name: 'Bienes de Consumo',
//             type: 'link'
//           },
//           /*          {
//                       state: 'cocina',
//                       name: 'Equipo de Cocina',
//                       type: 'link'
//                     },*/
//           /*          {
//                       state: 'medico',
//                       name: 'Equipo Medico',
//                       type: 'link'
//                     },*/
//           /*          {
//                       state: 'instrumental',
//                       name: 'Instrumental',
//                       type: 'link'
//                     },*/
//         ]
//       },
//       {state: 'seguimiento', name: 'Seguimiento', type: 'link'},
//       {state: 'editar', name: 'Editar', type: 'link'}
//     ]
//   },
//   {
//     state: 'configuraciones',
//     name: 'Configuraciones',
//     type: 'sub',
//     icon: 'build',
//     permiso: 'Jefe Biomedico',
//
//     children: [
//       {state: 'fecha', name: 'Fecha Límite', type: 'link'},
//       {state: 'presupuesto', name: 'Presupuesto', type: 'link'}
//
//     ]
//   },
//   {
//     state: 'usuarios',
//     name: 'Usuarios',
//     type: 'sub',
//     icon: 'supervised_user_circle',
//     permiso: 'Jefe Biomedico',
//
//     children: [
//       {state: 'validar', name: 'Validar Usuarios', type: 'link'},
//       {state: 'bloquear', name: 'Bloquear', type: 'link'}
//
//     ]
//   },
//   {
//     state: 'requerimientos',
//     name: 'Validar Solicitudes',
//     type: 'sub',
//     icon: 'assignment_turned_in',
//     permiso: 'Jefe Biomedico',
//
//     children: [
//       {state: 'revision', name: 'Validar Solicitud', type: 'link'},
//       {state: 'validado', name: 'Solicitudes Aprobadas', type: 'link'},
//       {state: 'final', name: 'Solicitud Final', type: 'link'}
//
//     ]
//   },
//   {
//     state: 'proveedores',
//     name: 'Proveedores',
//     type: 'sub',
//     icon: 'apartment',
//     permiso: 'Consulta',
//
//     children: [
//       {state: 'lista', name: 'Lista de Proveedores', type: 'link'},
//     ]
//   },
//   {
//     state: 'contratos',
//     name: 'Contratos',
//     type: 'sub',
//     icon: 'receipt',
//     permiso: 'Administrador de contratos',
//
//     children: [
//       {state: 'administrar', name: 'Administrar', type: 'link'},
//       {state: 'seguimiento', name: 'Seguimiento', type: 'link'},
//     ]
//   },
//   {
//     state: 'traslado',
//     name: 'Traslados de pacientes',
//     type: 'sub',
//     icon: 'receipt',
//     permiso: 'Solicitud traslado',
//
//     children: [
//       {state: 'solicitud', name: 'Nueva Solicitud', type: 'link'},
//       {state: 'realizadas', name: 'Solicitudes Realizadas', type: 'link'},
//       {state: 'cancelar', name: 'Cancelar Solicitud', type: 'link'},
//     ]
//   },
// ];

@Injectable()
export class MenuItems {
  // public men: any[];
  // public men1: any[];
  // public men2: any[];
  // public men3: any[];
  // public men4: any[];
  // public permisos: any[];
  public MENUITEM = [];

  constructor() {
  }


  getMenuitems(): Menu[] {
    return this.MENUITEM;
  }

  // getMenuitem(): Menu[] {
  //   /* ******  funcion que concatena los menus por default quue tienen todos los usuarios  ******   */
  //   this.men = MENUITEMS.filter(menu => {
  //     return menu.permiso === 'Consulta';
  //   });
  //   for (let i of this.permisos) {
  //     if (i === 'Administrador de contratos') {
  //       this.men2 = MENUITEMS.filter(menu => {
  //         return menu.permiso === 'Administrador de contratos';
  //       });
  //       this.men = this.men.concat(this.men2);
  //     }
  //     if (i === 'Jefe Biomedico') {
  //       this.men1 = MENUITEMS.filter(menu => {
  //         return menu.permiso === i;
  //       });
  //       this.men = this.men.concat(this.men1);
  //     }
  //     if (i === 'Solicitud traslado') {
  //       this.men4 = MENUITEMS.filter(menu => {
  //         return menu.permiso === i;
  //       });
  //       this.men = this.men.concat(this.men4);
  //     }
  //     if (i === 'Solicitud') {
  //       this.men3 = MENUITEMS.filter(menu => {
  //         return menu.permiso === i;
  //       });
  //       this.men = this.men.concat(this.men3);
  //     }
  //   }
  //   /* ******  regreso el menu  ******   */
  //   const tem = new Set(this.men);
  //   this.men = Array.from(tem);
  //   return this.men;
  // }
}
